protostar/commands/test/test_command.py

```diff
@@ -38,6 +40,58 @@
 from .test_command_cache import TestCommandCache
 
 
+@dataclass
+class SuccessfulTestMessage(StructuredMessage):
+    summary: TestingSummary
+    test_collector_result: TestCollector.Result
+
+    def format_human(self, fmt: LogColorProvider) -> str:
+        assert False, "Tests should use live logging for the human-readable output"
+
+    def format_dict(self) -> dict:
+        test_cases = 0
+        for test_suite in self.test_collector_result.test_suites:
+            test_cases += len(test_suite.test_cases)
+        failed_tests = len(self.summary.failed)
+        passed_tests = len(self.summary.passed)
+        execution_times = [
+            test.execution_time for test in self.summary.passed + self.summary.failed
+        ]
+
+        failed_tests_paths = {str(item.file_path) for item in self.summary.failed}
+        passed_test_suites = failed_test_suites = 0
+        for test_suite in self.test_collector_result.test_suites:
+            if str(test_suite.test_path) in failed_tests_paths:
+                failed_test_suites += 1
+            else:
+                passed_test_suites += 1
+
+        return {
+            "collected_tests": {
```

Comment:
I think `TestCollectingResult` deserves the own `Message` class.

Reply:
I'm not sure what's the gain here. Also, should we create a separate class for the summary as well? This looks pretty short and I'm not sure it's going to be reused.

Reply:
Protostar shows collected tests information before running tests, unlike the testing summary. The JSON argument shouldn't change messages order. It's confusing for users and developers.

Protostar might be converted to a plugin of the CargoLikeTool and JSON format could be used to communicate with Protostar, so messages shouldn't be delayed if it can be avoided.

> should we create a separate class for the summary as wel

This data class can be converted to represent the summary.

Reply:
Discussed offline